Spigot-Server-Patches/0441-Configurable-projectile-relative-velocity.patch

```diff
@@ -0,0 +1,89 @@
+
+This patch adds an option "disable relative projectile velocity", which, when
+nabled, will cause projectiles to ignore the shooter's current velocity,
+like they did in Minecraft 1.8 and prior.
+If a player is falling, for example, their shooting range will be drastically
+reduced, as a downwards velocity is applied to the projectile. This prevents
+players from saving themselves from falling off floating islands, for example,
+as a thrown ender pearl will not make it back to the island, while it would
+have in 1.8.
+
+While this could easily be done with plugins, too, there are multiple problems:
+P1) If multiple plugins cancel the velocity by subtracting the shooter's velocity
+from the projectile's velocity, the projectile's velocity would be different.
+As there's no way to detect whether the projectile's velocity has already been
+adjusted to ignore the player's velocity, plugins can't not do it if it's not
+necessary.
+P2) I've noticed some inconsistencies, e.g. weird velocity when shooting while
+using an elytra. Checking for those inconsistencies is possible, but not as
+efficient as just not applying the velocity in the first place.
+P3) Solutions for 1) and especially 2) might not be future-proof, while this
+server-internal fix makes this change future-proof.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index bfd690ec..b8131aba 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -474,6 +474,11 @@ public class PaperWorldConfig {
+         log("Disable Unloaded Chunk Enderpearl Exploit: " + (disableEnderpearlExploit ? "enabled" : "disabled"));
+     }
+ 
++    public boolean disableRelativeProjectileVelocity;
++    private void disableRelativeProjectileVelocity() {
++        disableRelativeProjectileVelocity = getBoolean("game-mechanics.disable-relative-projectile-velocity", false);
++    }
++
+     public boolean villagesLoadChunks = false;
+     private void villagesLoadChunks() {
+         villagesLoadChunks = getBoolean("game-mechanics.villages-load-chunks", false);
+diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
+index 6aa518f1..0fee2b00 100644
+--- a/src/main/java/net/minecraft/server/EntityArrow.java
++++ b/src/main/java/net/minecraft/server/EntityArrow.java
+@@ -79,13 +79,17 @@ public abstract class EntityArrow extends Entity implements IProjectile {
+         float f6 = -MathHelper.sin(f * 0.017453292F);
+         float f7 = MathHelper.cos(f1 * 0.017453292F) * MathHelper.cos(f * 0.017453292F);
+ 
++
+         this.shoot((double) f5, (double) f6, (double) f7, f3, f4);
++        // Paper start - if "disable relative projectile velocity" is active, enforce it
++        if(!entity.world.paperConfig.disableRelativeProjectileVelocity) {
```

Comment:
For future reference, there should be a space here. `if (!entity.world.etc...`
This is super minor so not something we'd ask you to do, we can make this super tiny change at merge.